File: bindings/typescript/src/workflows.js
import { RustCrateBridge } from "./bridge.js";

function mergeState(left, right) {
  return { ...left, ...right };
}

export class TokenBudget {
  constructor(maxContextTokens, maxCompletionTokens) {
    this.maxContextTokens = maxContextTokens;
    this.maxCompletionTokens = maxCompletionTokens;
  }

  fits(promptTokens, completionTokens) {
    return promptTokens + completionTokens <= this.maxContextTokens;
  }
}

export class RetryPolicy {
  constructor(maxAttempts, baseDelayMs, maxDelayMs, jitter = true) {
    this.maxAttempts = maxAttempts;
    this.baseDelayMs = baseDelayMs;
    this.maxDelayMs = maxDelayMs;
    this.jitter = jitter;
  }
}

export class CoreOrchestrator {
  enforceCompletionBudget(budget, promptTokens) {
    if (!budget.fits(promptTokens, budget.maxCompletionTokens)) {
      throw new Error("budget exceeded");
    }
  }

  nextRetryDelay(policy, attempt) {
    if (attempt <= 0 || attempt > policy.maxAttempts) {
      throw new Error(`invalid retry attempt: ${attempt}`);
    }
    const base = Math.min(
      policy.baseDelayMs * (2 ** (attempt - 1)),
      policy.maxDelayMs,
    );
    if (!policy.jitter || base === 0) return base;
    return Math.floor(Math.random() * (base + 1));
  }
}

export class PrismConfig {
  constructor(otlpEndpoint, serviceName = "orchustr-typescript") {
    this.otlpEndpoint = otlpEndpoint;
    this.serviceName = serviceName;
  }
}

export function installGlobalSubscriber(otlpEndpoint) {
  return RustCrateBridge.invoke("or-prism", "install_global_subscriber", {
    otlp_endpoint: otlpEndpoint,
  });
}

export class PlainText {
  constructor(text) {
    this.text = text;
  }
}

export class TextParser {
  parse(raw) {
    const text = raw.trim();
    if (!text) throw new Error("text must not be empty");
    return new PlainText(text);
  }
}

export class CheckpointGate {
  constructor() {
    this.records = new Map();
  }

  async pause(checkpointId, resumeFrom, state) {
    this.records.set(checkpointId, {
      checkpointId,
      resumeFrom,
      state: { ...state },
    });
  }

  async resume(checkpointId) {
    if (!this.records.has(checkpointId)) {
      throw new Error(`unknown checkpoint: ${checkpointId}`);
    }
    return this.records.get(checkpointId);
  }
}

export class RecallEntry {
  constructor(kind, value) {
    this.kind = kind;
    this.value = value;
  }
}

export class RecallStore {
  constructor() {
    this.entries = [];
  }

  async store(entry) {
    this.entries.push(entry);
  }

  async list(kind) {
    return this.entries.filter((entry) => entry.kind === kind);
  }
}

export class RecallOrchestrator {
  async remember(store, entry) {
    await store.store(entry);
  }

  async recall(store, kind) {
    return await store.list(kind);
  }
}

export class CompassRouterBuilder {
  constructor() {
    this.routes = [];
    this.defaultRoute = null;
  }

  addRoute(name, predicate) {
    this.routes.push([name, predicate]);
    return this;
  }

  setDefault(route) {
    this.defaultRoute = route;
    return this;
  }

  build() {
    const routes = [...this.routes];
    const defaultRoute = this.defaultRoute;
    return {
      select(state) {
        for (const [name, predicate] of routes) {
          if (predicate(state)) return { route: name };
        }
        if (defaultRoute) return { route: defaultRoute };
        throw new Error("no matching route");
      },
    };
  }
}

export class PipelineBuilder {
  constructor() {
    this.nodes = [];
  }

  addNode(name, handler) {
    this.nodes.push([name, handler]);
    return this;
  }

  add_node(name, handler) {
    return this.addNode(name, handler);
  }

  build() {
    if (this.nodes.length === 0) {
      throw new Error("pipeline requires at least one node");
    }
    const nodes = [...this.nodes];
    return {
      async execute(initialState) {
        let state = { ...initialState };
        for (const [, handler] of nodes) {
          state = mergeState(state, { ...(await handler({ ...state })) });
        }
        return state;
      },
      async invoke(initialState) {
        return this.execute(initialState);
      },
    };
  }
}

export class RelayBuilder {
  constructor() {
    this.branches = [];
  }

  addBranch(name, handler) {
    this.branches.push([name, handler]);
    return this;
  }

  add_branch(name, handler) {
    return this.addBranch(name, handler);
  }

  build() {
    if (this.branches.length === 0) {
      throw new Error("relay requires at least one branch");
    }
    return { branches: [...this.branches] };
  }
}

export class RelayExecutor {
  async execute(plan, initialState) {
    const patches = await Promise.all(
      plan.branches.map(async ([name, handler]) => [
        name,
        { ...(await handler({ ...initialState })) },
      ]),
    );
    let state = { ...initialState };
    for (const [, patch] of patches.sort((a, b) => a[0].localeCompare(b[0]))) {
      state = mergeState(state, patch);
    }
    return state;
  }
}

export class ColonyOrchestrator {
  constructor() {
    this.members = [];
  }

  addMember(name, role, agent) {
    this.members.push([{ name, role }, agent]);
    return this;
  }

  async coordinate(initialState) {
    if (this.members.length === 0) {
      throw new Error("colony requires at least one member");
    }
    const transcript = [];
    let state = { ...initialState };
    for (const [member, agent] of this.members) {
      const response = await agent({ ...state }, [...transcript], member);
      const message = response && typeof response === "object" && "content" in response
        ? response
        : { from: member.name, to: "all", content: String(response) };
      transcript.push(message);
      state = { ...state, [member.name]: message.content };
    }
    return {
      summary: transcript.at(-1)?.content ?? "",
      state,
      transcript,
    };
  }
}

export class ColonyBuilder {
  constructor() {
    this.orchestrator = new ColonyOrchestrator();
  }

  addMember(name, role, agent) {
    this.orchestrator.addMember(name, role, agent);
    return this;
  }

  add_member(name, role, agent) {
    return this.addMember(name, role, agent);
  }

  build() {
    return this.orchestrator;
  }
}

export class SentinelConfig {
  constructor(maxSteps = 8, metadata = {}) {
    this.maxSteps = maxSteps;
    this.metadata = metadata;
  }
}

export class StepOutcome {
  constructor(status, state, message = null) {
    this.status = status;
    this.state = state;
    this.message = message;
  }
}

export class SentinelOrchestrator {
  async runAgent(agent, initialState, config) {
    const result = await agent({ ...initialState }, config);
    if (result instanceof StepOutcome) return result;
    if (typeof result === "object" && result !== null) {
      return new StepOutcome("completed", result);
    }
    return new StepOutcome("completed", { ...initialState }, String(result));
  }
}

File: bindings/dart/lib/src/workflows.dart
import "dart:async";
import "dart:math";

import "bridge.dart";
import "types.dart";

JsonObject _merge(JsonObject left, JsonObject right) =>
    <String, Object?>{...left, ...right};

Future<JsonObject> _toJsonObject(FutureOr<JsonObject> value) async =>
    copyJsonObject(await Future.value(value));

final class TokenBudget {
  const TokenBudget(this.maxContextTokens, this.maxCompletionTokens);

  final int maxContextTokens;
  final int maxCompletionTokens;

  bool fits(int promptTokens, int completionTokens) {
    return promptTokens + completionTokens <= maxContextTokens;
  }
}

final class RetryPolicy {
  const RetryPolicy(this.maxAttempts, this.baseDelayMs, this.maxDelayMs,
      {this.jitter = true});

  final int maxAttempts;
  final int baseDelayMs;
  final int maxDelayMs;
  final bool jitter;
}

final class CoreOrchestrator {
  void enforceCompletionBudget(TokenBudget budget, int promptTokens) {
    if (!budget.fits(promptTokens, budget.maxCompletionTokens)) {
      throw StateError("budget exceeded");
    }
  }

  int nextRetryDelay(RetryPolicy policy, int attempt) {
    if (attempt <= 0 || attempt > policy.maxAttempts) {
      throw StateError("invalid retry attempt: $attempt");
    }
    final base =
        min(policy.baseDelayMs * (1 << (attempt - 1)), policy.maxDelayMs);
    if (!policy.jitter || base == 0) {
      return base;
    }
    return Random().nextInt(base + 1);
  }
}

final class PrismConfig {
  const PrismConfig(this.otlpEndpoint, {this.serviceName = "orchustr-dart"});

  final String otlpEndpoint;
  final String serviceName;
}

Object? installGlobalSubscriber(String otlpEndpoint) {
  return RustCrateBridge.invoke("or-prism", "install_global_subscriber",
      <String, Object?>{"otlp_endpoint": otlpEndpoint});
}

final class PlainText {
  const PlainText(this.text);

  final String text;
}

final class TextParser {
  PlainText parse(String raw) {
    final text = raw.trim();
    if (text.isEmpty) {
      throw StateError("text must not be empty");
    }
    return PlainText(text);
  }
}

final class CheckpointRecord {
  const CheckpointRecord(this.checkpointId, this.resumeFrom, this.state);

  final String checkpointId;
  final String resumeFrom;
  final JsonObject state;
}

final class CheckpointGate {
  final Map<String, CheckpointRecord> _records = <String, CheckpointRecord>{};

  Future<void> pause(
      String checkpointId, String resumeFrom, JsonObject state) async {
    _records[checkpointId] =
        CheckpointRecord(checkpointId, resumeFrom, copyJsonObject(state));
  }

  Future<CheckpointRecord> resume(String checkpointId) async {
    final record = _records[checkpointId];
    if (record == null) {
      throw StateError("unknown checkpoint: $checkpointId");
    }
    return record;
  }
}

final class RecallEntry {
  const RecallEntry(this.kind, this.value);

  final String kind;
  final JsonObject value;
}

final class RecallStore {
  final List<RecallEntry> _entries = <RecallEntry>[];

  Future<void> store(RecallEntry entry) async {
    _entries.add(entry);
  }

  Future<List<RecallEntry>> list(String kind) async {
    return _entries.where((RecallEntry entry) => entry.kind == kind).toList();
  }
}

final class RecallOrchestrator {
  Future<void> remember(RecallStore store, RecallEntry entry) {
    return store.store(entry);
  }

  Future<List<RecallEntry>> recall(RecallStore store, String kind) {
    return store.list(kind);
  }
}

typedef RoutePredicate = bool Function(JsonObject state);

final class CompassRouterBuilder {
  final List<(String, RoutePredicate)> _routes = <(String, RoutePredicate)>[];
  String? _defaultRoute;

  CompassRouterBuilder addRoute(String name, RoutePredicate predicate) {
    _routes.add((name, predicate));
    return this;
  }

  CompassRouterBuilder setDefault(String route) {
    _defaultRoute = route;
    return this;
  }

  CompassRouter build() => CompassRouter(List<(String, RoutePredicate)>.from(_routes), _defaultRoute);
}

final class CompassRouter {
  const CompassRouter(this._routes, this._defaultRoute);

  final List<(String, RoutePredicate)> _routes;
  final String? _defaultRoute;

  String select(JsonObject state) {
    for (final (String route, RoutePredicate predicate) in _routes) {
      if (predicate(state)) {
        return route;
      }
    }
    final fallback = _defaultRoute;
    if (fallback == null) {
      throw StateError("no matching route");
    }
    return fallback;
  }
}

typedef PipelineNode = FutureOr<JsonObject> Function(JsonObject state);

final class PipelineBuilder {
  final List<(String, PipelineNode)> _nodes = <(String, PipelineNode)>[];

  PipelineBuilder addNode(String name, PipelineNode handler) {
    _nodes.add((name, handler));
    return this;
  }

  Pipeline build() {
    if (_nodes.isEmpty) {
      throw StateError("pipeline requires at least one node");
    }
    return Pipeline(List<(String, PipelineNode)>.from(_nodes));
  }
}

final class Pipeline {
  const Pipeline(this._nodes);

  final List<(String, PipelineNode)> _nodes;

  Future<JsonObject> execute(JsonObject initialState) async {
    var state = copyJsonObject(initialState);
    for (final (_, PipelineNode handler) in _nodes) {
      state = _merge(state, await _toJsonObject(handler(copyJsonObject(state))));
    }
    return state;
  }
}

typedef RelayBranch = FutureOr<JsonObject> Function(JsonObject state);

final class RelayBuilder {
  final List<(String, RelayBranch)> _branches = <(String, RelayBranch)>[];

  RelayBuilder addBranch(String name, RelayBranch handler) {
    _branches.add((name, handler));
    return this;
  }

  RelayPlan build() {
    if (_branches.isEmpty) {
      throw StateError("relay requires at least one branch");
    }
    return RelayPlan(List<(String, RelayBranch)>.from(_branches));
  }
}

final class RelayPlan {
  const RelayPlan(this.branches);

  final List<(String, RelayBranch)> branches;
}

final class RelayExecutor {
  Future<JsonObject> execute(RelayPlan plan, JsonObject initialState) async {
    final patches = await Future.wait(plan.branches.map(((
      String,
      RelayBranch,
    ) branch) async {
      final (String name, RelayBranch handler) = branch;
      return (name, await _toJsonObject(handler(copyJsonObject(initialState))));
    }));
    patches.sort((left, right) => left.$1.compareTo(right.$1));
    var state = copyJsonObject(initialState);
    for (final (_, JsonObject patch) in patches) {
      state = _merge(state, patch);
    }
    return state;
  }
}

typedef ColonyAgent = FutureOr<Object?> Function(
  JsonObject state,
  List<ColonyMessage> transcript,
  ColonyMember member,
);

final class ColonyMember {
  const ColonyMember(this.name, this.role);

  final String name;
  final String role;
}

final class ColonyMessage {
  const ColonyMessage(this.from, this.to, this.content);

  final String from;
  final String to;
  final String content;
}

final class ColonyResult {
  const ColonyResult(this.summary, this.state, this.transcript);

  final String summary;
  final JsonObject state;
  final List<ColonyMessage> transcript;
}

final class ColonyOrchestrator {
  final List<(ColonyMember, ColonyAgent)> _members =
      <(ColonyMember, ColonyAgent)>[];

  ColonyOrchestrator addMember(String name, String role, ColonyAgent agent) {
    _members.add((ColonyMember(name, role), agent));
    return this;
  }

  Future<ColonyResult> coordinate(JsonObject initialState) async {
    if (_members.isEmpty) {
      throw StateError("colony requires at least one member");
    }
    var state = copyJsonObject(initialState);
    final List<ColonyMessage> transcript = <ColonyMessage>[];
    for (final (ColonyMember member, ColonyAgent agent) in _members) {
      final response = await Future.value(
        agent(copyJsonObject(state), List<ColonyMessage>.from(transcript), member),
      );
      final message = switch (response) {
        ColonyMessage value => value,
        _ => ColonyMessage(member.name, "all", "$response"),
      };
      transcript.add(message);
      state[member.name] = message.content;
    }
    return ColonyResult(
      transcript.isEmpty ? "" : transcript.last.content,
      state,
      transcript,
    );
  }
}

final class SentinelConfig {
  const SentinelConfig({this.maxSteps = 8, this.metadata = const <String, Object?>{}});

  final int maxSteps;
  final JsonObject metadata;
}

final class StepOutcome {
  const StepOutcome(this.status, this.state, {this.message});

  final String status;
  final JsonObject state;
  final String? message;
}

typedef SentinelAgent = FutureOr<Object?> Function(
  JsonObject state,
  SentinelConfig config,
);

final class SentinelOrchestrator {
  Future<StepOutcome> runAgent(
    SentinelAgent agent,
    JsonObject initialState,
    SentinelConfig config,
  ) async {
    final result = await Future.value(agent(copyJsonObject(initialState), config));
    return switch (result) {
      StepOutcome value => value,
      JsonObject state => StepOutcome("completed", state),
      _ => StepOutcome("completed", initialState, message: "$result"),
    };
  }
}
